authentication/oidc.go

```diff
@@ -434,3 +434,46 @@ func (a oidcAuthenticator) checkAuth(ctx context.Context, token string) (context
 
 	return ctx, "", http.StatusOK, codes.OK
 }
+
+// EnforceAccessTokenPresentOnSignalWrite enforces that the Authorization header is present in the incoming request
+// for the given list of tenants. Otherwise, it returns an error.
+// It protects the Prometheus remote write and Loki push endpoints. The tracing endpoint is not protected because
+// it goes through the gRPC middleware stack, which behaves differently from the HTTP one.
+func EnforceAccessTokenPresentOnSignalWrite(oidcTenants map[string]struct{}) Middleware {
+	protectedSignalWritePath := []string{
+		// Prometheus remote write endpoint
+		"/api/v1/receive",
```

Comment:
This looks to me like we are crossing abstraction boundaries. I'm worried about hardcoding API endpoints for data into authentication middleware. The auth middleware should really not know anything about this. Imagine we one day change an endpoint or add a new one, now we have to somehow remember to update this list as well. I would worry this is asking for vulnerability to be introduced.

Maybe we could get around this by having a middleware factory that accepts a list of endpoints and we can produce this middleware in the main.go.

Reply:
> This looks to me like we are crossing abstraction boundaries.
I had a similar feeling about it, @squat. I tried to think of a way to prepend (not append, because we need this to run before the auth middleware, as it will do the redirect) this middleware to the stack of each sub-router (for metrics, logs, and traces... in their respective blocks in `main.go`). I didn't find a good way of doing this though. I'll investigate this more.

I was even wondering whether we could "cut the problem at its root" and remove the login redirect in the OIDC authentication flow and let it a responsibility of the clients. What do you think?